pkg/services/alerting/engine.go

```diff
@@ -86,32 +87,79 @@ func (e *Engine) runJobDispatcher(grafanaCtx context.Context) error {
 		case <-grafanaCtx.Done():
 			return dispatcherGroup.Wait()
 		case job := <-e.execQueue:
-			dispatcherGroup.Go(func() error { return e.processJob(alertCtx, job) })
+			dispatcherGroup.Go(func() error { return e.processJobWithRetry(alertCtx, job) })
 		}
 	}
 }
 
 var (
 	unfinishedWorkTimeout time.Duration = time.Second * 5
 	alertTimeout          time.Duration = time.Second * 30
+	alertMaxAttempts      int           = 2
 )
 
-func (e *Engine) processJob(grafanaCtx context.Context, job *Job) error {
+func (e *Engine) processJobWithRetry(grafanaCtx context.Context, job *Job) error {
+	defer func() {
+		if err := recover(); err != nil {
+			e.log.Error("Alert Panic", "error", err, "stack", log.Stack(1))
+		}
+	}()
+
+	cancelChan := make(chan context.CancelFunc, alertMaxAttempts)
+	attemptChan := make(chan int, 1)
+
+	// Initialize with first attemptID=1
+	attemptChan <- 1
+	job.Running = true
+
+	for {
+		select {
+		case <-grafanaCtx.Done():
+			unfinishedWorkTimer := time.NewTimer(unfinishedWorkTimeout)
+			for {
```

Comment:
I think it should be possible to have only one for { select { } } instead of two, and that would make the code more readable

Reply:
I looks like it's not possible. I mean, it would be possible to have something that works but not as expected.
It would looks like 
```
	for {
		select {
		case <-grafanaCtx.Done():
			time.AfterFunc(unfinishedWorkTimeout, func() {
				endChan <- grafanaCtx.Err()
			})
		case err := <-endChan:
			return e.endJob(err, cancelChan, job)
		case attemptID, more := <-attemptChan:
			if !more {
				return e.endJob(nil, cancelChan, job)
			}
			go e.processJob(attemptID, attemptChan, cancelChan, job)
		}
	}
```

In golang `select` chooses one at random if multiple cases are ready.
In addition, we do need to wait for `<-grafanaCtx.Done()` to be ready, to catch grafana server context cancellation. And `<-grafanaCtx.Done()` will remain a valid case, even after executing that case.
Therefore it will loop on execution of `<-grafanaCtx.Done()` case a lot of time between grafana server context cancellation and the end of job processing; plus a random times when multiple cases are ready.
That's not the behavior that we would like.

The core issue, is that AFAIK **there is no way in golang to prioritize a case in a `select` statement**. The common workarroud (as already used here before my change) seems to have a `select` inside a `select`. And because we can now wait for several attempts, it's `for + select` inside `for + select`

Reply:
Maybe it would be worth it putting the inner for in a separate method/function then.

While I understand that you want to let the job finish, why do we allow new attempts when `grafanaCtx.Done()` ? (What is `grafanaCtx.Done()` exactly ?)

Reply:
`grafanaCtx.Done()` gets a signal when Grafana wants to shut down. Which gives alerting some time to save its current state to the database first. 

It's slight abusive way of using Context. But back when I wrote it I didn't find any other solution that I liked.

Reply:
Ok, so in this case clearly we don't want to attempt new retries on grafanaCtx.Done() :)

Reply:
Fixed